### PR TITLE
Add `ffi_call` tutorial

### PR DESCRIPTION
**Pull request by dfm**
Building on https://github.com/google/jax/pull/21925, this tutorial demonstrates the use of the FFI using `ffi_call` with a simple example. I don't think this should cover all of the most advanced use cases, but it should be sufficient for the most common examples. I think it would be useful to eventually replace the existing CUDA tutorial, but I'm not sure that it'll get there in the first draft.

As an added benefit, this also runs a simple test (akin to `docs/cuda_custom_call`) which actually executes using a tool chain that open source users would use in practice.

The current rendered version of the tutorial is here: https://jax--22095.org.readthedocs.build/en/22095/ffi.html

### CODE REVIEW

File: docs/ffi/ffi.md [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+!cmake --build _build
+!cmake --install _build
+```
+
+With this compiled library in hand, we now need to register this handler with XLA via the {func}`~jax.lib.xla_client.register_custom_call_target` function.
+In our nanobind module above, we implemented a Python function called `rms_norm` which returns a [`PyCapsule`](https://docs.python.org/3/c-api/capsule.html) containing a function pointer to the C++ function `RmsNorm`.
+This "capsule" is the object that we pass to `register_custom_call_target`:
+
+```{code-cell} ipython3
+from jax.lib import xla_client
```
Comment (hawkinsp) [inline docs/ffi/ffi.md:187]: Can we expose a public name for this in the jex.ffi namespace?

I'd like users not to need to know about `xla_client`.

File: docs/ffi/ffi.md [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+#include "xla/ffi/api/c_api.h"
+#include "xla/ffi/api/ffi.h"
+
+namespace ffi = xla::ffi;
+
+std::pair<int64_t, int64_t> GetDims(ffi::Span<const int64_t> dims) {
+  if (dims.size() == 0) {
+    return std::make_pair(0, 0);
+  }
+  int64_t totalSize =
```
Comment (superbobry) [inline docs/ffi/ffi.md:109]: I think this is `ffi::Buffer::element_count()`.

If not, it is probably worth adding new APIs to `ffi::Buffer` if they are generally useful.

File: docs/ffi/ffi.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+```c++
+#include <type_traits>
+
+#include "nanobind/nanobind.h"
+#include "xla/ffi/api/c_api.h"
+
+namespace nb = nanobind;
+
+template <typename T>
+nb::capsule EncapsulateFfiCall(T *fn) {
```
Comment (superbobry) [inline docs/ffi/ffi.md:159]: Should we bundle this helper somewhere in jaxlib? We can implement it using Python C API without depending on nanobind.

File: docs/ffi/ffi.md [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+```{code-cell} ipython3
+from jax.lib import xla_client
+import rms_norm as rms_norm_lib
+
+xla_client.register_custom_call_target(
+  "rms_norm", rms_norm_lib.rms_norm(), platform="cpu", api_version=1
+)
+```
+
+We note that the `api_version=1` keyword in the call to `register_custom_call_target` is important because it indicates that our function pointer implements this FFI interface.
```
Comment (superbobry) [inline docs/ffi/ffi.md:195]: s/this/typed/?

Alternatively, we can say that `api_version=0` is the legacy FFI interface which we are not covering in the doc?

File: docs/ffi/ffi.md [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+
+by the time the function is compiled, the appropriate FFI has been selected:
+
+```{code-cell} ipython3
+print(jax.jit(rms_norm_cross_platform).lower(x).compile().as_text().strip())
+```
+
+and there will be no runtime overhead to using {func}`jax.lax.platform_dependent`.
+
+## Advanced topics
```
Comment (superbobry) [inline docs/ffi/ffi.md:460]: Maybe mention stateful custom calls as well?

**Review by superbobry — APPROVED** [review]

File: docs/ffi/ffi.md (hunk 3 above)
Comment (dfm) [reply docs/ffi/ffi.md:159]: This is an interesting idea! My first reaction is that it doesn't really make sense to include this in `jaxlib` since it's an implementation detail and users could choose to use `ctypes` and `jex.ffi.pycapsule` if they prefer (this reminds me that I also meant to add a discussion of that to this tutorial). But, it would probably improve the developer experience if we have a clearer recommended approach. I'm inclined to leave that to future work rather than trying to overcommit with this tutorial for now.

File: docs/ffi/ffi.md [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+  return scale;
+}
+```
+
+and, for our example, this is the function that we want to expose to JAX via the FFI.
+
+### C++ interface
+
+To expose our library function to JAX and XLA, we need to write a thin wrapper using the APIs provided by the header-only library in the [`xla/ffi/api`](https://github.com/openxla/xla/tree/main/xla/ffi/api) directory of the [XLA project](https://github.com/openxla/xla).
+For more information about this interface, take a look at [the XLA custom call documentation](https://openxla.org/xla/custom_call).
```
Comment (gnecula) [inline docs/ffi/ffi.md:92]: Is there documentation for the `Ffi::Ffi::Bind`? Last time I looked a few months ago, I could not find any.

File: docs/ffi/ffi.md [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+#include <functional>
+#include <numeric>
+#include <utility>
+
+#include "xla/ffi/api/c_api.h"
+#include "xla/ffi/api/ffi.h"
+
+namespace ffi = xla::ffi;
+
+std::pair<int64_t, int64_t> GetDims(ffi::Span<const int64_t> dims) {
```
Comment (gnecula) [inline docs/ffi/ffi.md:105]: Please consider adding some comments in the code, as you do below in the Python file.

File: docs/ffi/ffi.md [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+        .Attr<float>("eps")
+        .Arg<ffi::Buffer<ffi::DataType::F32>>(/* x */)
+        .Ret<ffi::Buffer<ffi::DataType::F32>>(/* y */));
+```
+
+Starting at the bottom, we're using the XLA-provided macro `XLA_FFI_DEFINE_HANDLER` to generate some boilerplate which will expand into a function called `RmsNorm` with the appropriate signature.
+But, the important stuff here is all in the call to `ffi::Ffi::Bind()`, where we define the input and output types, and the types of any parameters.
+
+Then, in `RmsNormImpl`, we accept `ffi::Buffer` arguments which include information about the buffer shape, and pointers to the underlying data.
+In this implementation, we treat all leading dimensions of the buffer as batch dimensions, and perform RMS normalization over the last axis.
```
Comment (gnecula) [inline docs/ffi/ffi.md:140]: I find "all leading dimensions" to be ambiguous. For some kernels all but the last 2 dimension should be the batch dimensions, and for others perhaps all but the last one.

File: docs/ffi/ffi.md [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+
+# Check that this gives the right answer when compared to the reference version
+ct_y = jnp.ones_like(x)
+np.testing.assert_allclose(
+  jax.vjp(rms_norm, x)[1](ct_y), jax.vjp(rms_norm_ref, x)[1](ct_y), rtol=1e-5
+)
+```
+
+At this point, we can use our new `rms_norm` function transparently for many JAX applications, and it will transform appropriately under the standard JAX function transformations like {func}`~jax.vmap` and {func}`~jax.grad`.
+One thing that this example doesn't support is forward-mode AD ({func}`jax.jvp`, for example) since {func}`~jax.custom_vjp` is restricted to reverse-mode.
```
Comment (gnecula) [inline docs/ffi/ffi.md:371]: I think that this also does not support higher-order AD, I think.

File: docs/ffi/ffi.md [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+JAX doesn't currently expose a public API for simultaneously customizing both forward-mode and reverse-mode AD, but such an API is on the roadmap, so please [open an issue](https://github.com/google/jax/issues) describing you use case if you hit this limitation in practice.
+
+## FFI calls on a GPU
+
+So far, we have been interfacing only with foreign functions running on the CPU, but JAX's FFI also supports calls to GPU code.
+Since this documentation page is automatically generated on a machine without access to a GPU, we can't execute any GPU-specific examples here, but we will go over the key points.
```
Comment (gnecula) [inline docs/ffi/ffi.md:377]: I thought that this would not be runnable under `pytest doctest`, as we do for other .md files, but now I am wondering what kind of automation do you have. If it is not automatic, it would be good to add some comments somewhere for how a JAX developer can refresh the documentation.

File: docs/ffi/ffi.md [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+
+```{code-cell} ipython3
+jax.make_jaxpr(rms_norm_cross_platform)(x)
+```
+
+```{code-cell} ipython3
+print(jax.jit(rms_norm_cross_platform).lower(x).as_text().strip())
+```
+
+by the time the function is compiled, the appropriate FFI has been selected:
```
Comment (gnecula) [inline docs/ffi/ffi.md:452]: You can show the `as_text(dialect="hlo")`, which shows the HLO passed to the compiler.

File: docs/ffi/ffi.md [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+
+```{code-cell} ipython3
+print(jax.jit(rms_norm_cross_platform).lower(x).compile().as_text().strip())
+```
+
+and there will be no runtime overhead to using {func}`jax.lax.platform_dependent`.
```
Comment (gnecula) [inline docs/ffi/ffi.md:458]: not only won't there be runtime overhead but there is no danger that the compiler is given an input that contains custom calls with unknown targets.

File: docs/ffi/ffi.md (hunk 10 above)
Comment (dfm) [reply docs/ffi/ffi.md:377]: `doctest` doesn't actually test most of the ipynb/md files in `docs` since AFAIK it only tests the specific syntax that includes `>>>`. But, like many of the other notebooks, this document is automatically executed on Read the Docs when generating the docs page and the custom calls get compiled there because RtDs comes with the all the necessary developer tools. There's nothing special about this tutorial compared to the others.

One nice benefit is that this means that this tutorial gives us a (non-blocking) CI test for the user version of the FFI API!

File: docs/ffi/ffi.md (hunk 5 above)
Comment (dfm) [reply docs/ffi/ffi.md:460]: I'm not totally sure what to say here. Do you think it's useful to just say that future versions of the API might support stateful custom calls? Perhaps it would be useful to give some general examples. Can you think of some common examples?

File: docs/ffi/ffi.md (hunk 9 above)
Comment (dfm) [reply docs/ffi/ffi.md:377]: Good point - I've added a comment about this.

File: docs/ffi/ffi.md (hunk 11 above)
Comment (dfm) [reply docs/ffi/ffi.md:461]: I didn't know that! Updated.

File: docs/ffi/ffi.md (hunk 12 above)
Comment (dfm) [reply docs/ffi/ffi.md]: I've added this comment as well.

File: docs/ffi/ffi.md (hunk 2 above)
Comment (dfm) [reply docs/ffi/ffi.md]: I've updated the tutorial to use a recent version of the FFI headers and left a TODO to return to bundled version after the next jaxlib release. (Interesting side note, I couldn't use the most recent version of the headers because they don't seem to be backwards compatible with the released jaxlib. It seems to have something to do with the addition of the `instantiate` stage... I will investigate!)

File: docs/ffi/ffi.md (hunk 10 above)
Comment (gnecula) [reply docs/ffi/ffi.md:386]: Nice. Non-blocking has the disadvantage that somebody has to watch for breakages after submit :-(

File: docs/ffi/rms_norm.cc [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    ComputeRmsNorm(eps, lastDim, &(x.typed_data()[n]), &(y->typed_data()[n]));
+  }
+  return ffi::Error::Success();
+}
+
+// Wrap `RmsNormImpl` and specify the interface to XLA.
+XLA_FFI_DEFINE_HANDLER(RmsNorm, RmsNormImpl,
+                       ffi::Ffi::Bind()
+                           .Attr<float>("eps")
+                           .Arg<ffi::Buffer<ffi::DataType::F32>>(/* x */)
```
Comment (ezhulenev) [inline docs/ffi/rms_norm.cc:78]: I find `/* x */` as fake a parameter a little confusing, I prefer this version:

```
.Arg<ffi::Buffer<ffi::DataType::F32>>()  // x
```

File: docs/ffi/ffi.md (hunk 5 above)
Comment (ezhulenev) [reply docs/ffi/ffi.md:469]: We have stateful example here: https://github.com/openxla/xla/blob/main/xla/service/gpu/custom_call_test.cc#L794-L845

I suspect that stateful custom calls are most useful "advanced" feature, not sure it should be in this tutorial, maybe in a separate one.

File: docs/ffi/ffi.md (hunk 6 above)
Comment (ezhulenev) [reply docs/ffi/ffi.md:92]: I'll add more details to https://openxla.org/xla/custom_call

File: docs/ffi/ffi.md (hunk 8 above)
Comment (dfm) [reply docs/ffi/ffi.md:150]: I agree! I've added a bit more discussion and a link to the `vmap` section below hoping that this can help clarify things.

File: docs/ffi/ffi.md (hunk 10 above)
Comment (dfm) [reply docs/ffi/ffi.md:391]: Yes! I am planning to add something like this more directly to CI, but it is somewhat involved so I haven't gotten to it yet. It's on the to do list!

File: docs/ffi/rms_norm.cc (hunk 13 above)
Comment (dfm) [reply docs/ffi/rms_norm.cc]: Thanks! I don't feel strongly either way. I've updated to this version.

File: docs/ffi/ffi.md (hunk 3 above)
Comment (erikfrey) [reply docs/ffi/ffi.md:171]: Drive-by comment: having one less dependency is attractive.  So out of curiosity, why not default to ctypes and leave a discussion of nanobind to future work?  What does nanobind offer here that ctypes doesn't?

Comment (dfm) [reply docs/ffi/ffi.md:171]: Good question! The main reason why I'm using the nanobind version here is that it lets importlib handle the logic for loading the appropriate library across platforms. I've always struggled to do this cleanly with ctypes, but it should certainly be possible!

Another minor point is that it is tricky (though not impossible) to use the ctypes approach inside of Google, and I couldn't come up with an approach that worked cleanly both internally and externally, but I'm mostly targeting external users here so it's not a huge constraint!

Comment (dfm) [reply docs/ffi/ffi.md:171]: Either way, I think it would be useful to present the alternatives and include a bit more discussion in this tutorial. I will do that - thanks!
